colexecbase: missing casts between floats

roachtest.sqlsmith/setup=rand-tables/setting=no-mutations [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3199213&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3199213&tab=artifacts#/sqlsmith/setup=rand-tables/setting=no-mutations) on master @ [5fd1dda1cde0a91643d39209c6458dcefcbfc180](https://github.com/cockroachdb/cockroach/commits/5fd1dda1cde0a91643d39209c6458dcefcbfc180):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=rand-tables/setting=no-mutations/run_1
	sqlsmith.go:217,sqlsmith.go:252,test_runner.go:765: error: pq: internal error: unexpected error from the vectorized engine: unhandled cast float -> float4
		stmt:
		SELECT
			e'II\x1e\x04':::STRING AS col_1705,
			B'100111110111010101000110011101110101111011010001' AS col_1706,
			tab_779.col3_18 AS col_1707,
			NULL AS col_1708
		FROM
			defaultdb.public.table3@[0] AS tab_779
			JOIN defaultdb.public.table2@primary AS tab_780 ON (tab_779.col3_13) = (tab_780.col2_11)
		ORDER BY
			tab_780.col2_0 DESC, tab_780.tableoid;
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
# From https://go.crdb.dev/p/roachstress, perhaps edited lightly.
caffeinate ./roachstress.sh sqlsmith/setup=rand-tables/setting=no-mutations
```
</p>
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=rand-tables/setting=no-mutations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/colexec/colexecbase/cast.eg.go
+++ b/pkg/sql/colexec/colexecbase/cast.eg.go
@@ -50,8 +50,11 @@ func GetCastOperator(
 	if fromType.Family() == types.UnknownFamily {
 		return &castOpNullAny{castOpBase: base}, nil
 	}
-	if toType.Identical(fromType) {
-		// We have an identity cast, so we use a custom identity cast operator.
+	if toType.Identical(fromType) || (fromType.Family() == types.FloatFamily && toType.Family() == types.FloatFamily) {
+		// We either have an identity cast or we have a cast between floats (and
+		// all floats are represented by float64 physically, so they are
+		// essentially identical too), so we use a custom identity cast
+		// operator.
 		return &castIdentityOp{castOpBase: base}, nil
 	}
 	isFromDatum := typeconv.TypeFamilyToCanonicalTypeFamily(fromType.Family()) == typeconv.DatumVecCanonicalTypeFamily
@@ -287,7 +290,7 @@ func IsCastSupported(fromType, toType *types.T) bool {
 	if fromType.Family() == types.UnknownFamily {
 		return true
 	}
-	if toType.Identical(fromType) {
+	if toType.Identical(fromType) || (fromType.Family() == types.FloatFamily && toType.Family() == types.FloatFamily) {
 		return true
 	}
 	isFromDatum := typeconv.TypeFamilyToCanonicalTypeFamily(fromType.Family()) == typeconv.DatumVecCanonicalTypeFamily

--- a/pkg/sql/colexec/colexecbase/cast_tmpl.go
+++ b/pkg/sql/colexec/colexecbase/cast_tmpl.go
@@ -84,8 +84,11 @@ func GetCastOperator(
 	if fromType.Family() == types.UnknownFamily {
 		return &castOpNullAny{castOpBase: base}, nil
 	}
-	if toType.Identical(fromType) {
-		// We have an identity cast, so we use a custom identity cast operator.
+	if toType.Identical(fromType) || (fromType.Family() == types.FloatFamily && toType.Family() == types.FloatFamily) {
+		// We either have an identity cast or we have a cast between floats (and
+		// all floats are represented by float64 physically, so they are
+		// essentially identical too), so we use a custom identity cast
+		// operator.
 		return &castIdentityOp{castOpBase: base}, nil
 	}
 	isFromDatum := typeconv.TypeFamilyToCanonicalTypeFamily(fromType.Family()) == typeconv.DatumVecCanonicalTypeFamily
@@ -139,7 +142,7 @@ func IsCastSupported(fromType, toType *types.T) bool {
 	if fromType.Family() == types.UnknownFamily {
 		return true
 	}
-	if toType.Identical(fromType) {
+	if toType.Identical(fromType) || (fromType.Family() == types.FloatFamily && toType.Family() == types.FloatFamily) {
 		return true
 	}
 	isFromDatum := typeconv.TypeFamilyToCanonicalTypeFamily(fromType.Family()) == typeconv.DatumVecCanonicalTypeFamily

--- a/pkg/sql/logictest/testdata/logic_test/vectorize
+++ b/pkg/sql/logictest/testdata/logic_test/vectorize
@@ -1231,3 +1231,21 @@ SELECT CASE WHEN _bool THEN _string ELSE _string END FROM t64793;
 4
 5
 NULL
+
+# Regression test for not supporting casts between FLOAT4 and FLOAT8 (which have
+# the same physical representation in the vectorized engine) (#67793).
+statement ok
+CREATE TABLE t67793 (_float4 FLOAT4, _float8 FLOAT8);
+INSERT INTO t67793 VALUES (1, 2), (2, 1);
+
+query RR rowsort
+SELECT t1._float8, t2._float4 FROM t67793 AS t1, t67793 AS t2 WHERE t1._float4 = t2._float8
+----
+1  1
+2  2
+
+query RR rowsort
+SELECT _float4::FLOAT8, _float8::FLOAT4 FROM t67793
+----
+1  2
+2  1

